[GeoMechanicsApplication] Add micro climate condition

rfaasse opened the pull request:
**📝 Description**
This PR adds a new condition for microclimate, including effects like solar radiation, air temperature and wind speed. The new class is based on the work done by @johnvanesch @mnabideltares. This PR also adds test cases (both on an integration and unit level).

avdg81 committed:
Added Mohamed's micro-climate condition "as-is"

In order to compile, I had to make the following changes:
- `T_condition.hpp` was renamed to `T_condition.h`.
- Base class `TCondition` was renamed to `GeoTCondition`.

avdg81 committed:
Made `CalculateLocalSystem` a `public` member

In the base class, `CalculateLocalSystem` is `public` as well, rather
than `protected`. By making it `public`, it becomes easier to unit test
the class.

avdg81 committed:
Made `Initialize` and `InitializeSolutionStep` `public` members

In the base class, these member functions are `public` rather than
`protected`. By making them `public`, it becomes easier to unit test the
class.

Added a first unit test that verifies no exception is thrown when a
micro-climate condition is initialized.

avdg81 committed:
Added a first unit test for `InitializeSolutionStep`

The test checks whether calling `InitializeSolutionStep`, given a
condition with dummy properties, does not throw an exception.

The unit test for `Initialize` now also uses a 2-noded condition.

avdg81 committed:
Registered micro-climate condition in the GeoMechanicsApplication

Copied the code from Mohamed's branch and made the following changes:
- Renamed the conditions by prepending "Geo" to the original names.
- Renamed the data members by adding "Geo" to their names.

avdg81 committed:
Let the model part create the micro-climate condition

Previously, in the test for `InitializeSolutionStep`, the micro-climate
condition was created directly. It is, however, more convenient to let
the model part take care of that.

avdg81 committed:
Refactored the unit test for `InitializeSolutionStep`

By extracting several helper functions, the test body has become a lot
more compact. Using this setup will make it easier to create similar
tests for different condition configurations.

avdg81 committed:
Added a unit test for 3-noded (line) micro-climate conditions

Test whether no errors occur when initializing the solution step.

avdg81 committed:
Added a unit test for 4-noded (line) micro-climate conditions

Test whether no errors occur when initializing the solution step.

avdg81 committed:
Added a unit test for 5-noded (line) micro-climate conditions

Test whether no errors occur when initializing the solution step.

avdg81 committed:
Generalized the way nodes are created for a micro-climate condition

Previously, all nodes were created on a line. This works for line
conditions, but not for triangular or quadrilateral ones. With these
changes we can now pass a function object that creates the nodes for us.
The unit tests have been adapted such that they now use this mechanism.

avdg81 committed:
Creating a condition for testing now requires the dimension size

To make the function that creates a micro-climate condition for testing
more generally applicable, it now has a parameter for passing the
dimension size.

Also added a unit test to verify that when initializing the solution
step of a micro-climate condition acting on a linear triangle executes
without raising any errors.

avdg81 committed:
Added a test for a condition acting on a quadratic triangle

The unit test verifies that when initializing the solution step of a
micro-climate condition acting on a quadratic triangle executes without
raising any errors.

The helper function that creates the nodes for a triangle has been
extended such that it can also create the mid-side nodes of a quadratic
triangle.

avdg81 committed:
Added a test for a condition acting on a linear quadrilateral

The unit test verifies that when initializing the solution step of a
micro-climate condition acting on a linear quadrilateral executes
without raising any errors.

avdg81 committed:
Added a test for a condition acting on a serendipity element

The unit test verifies that when initializing the solution step of a
micro-climate condition acting on a serendipity element executes without
raising any errors.

avdg81 committed:
Added a test for a condition acting on a LaGrange element

The unit test verifies that when initializing the solution step of a
micro-climate condition acting on a LaGrange element executes without
raising any errors.

avdg81 committed:
Added two unit tests for calculating the local system

The unit tests verify the left hand side matrix and the right hand side
vector calculated for a line micro-climate condition with two nodes.

To carry out the calculation, we also had to add solution step variables
AIR_HUMIDITY, PRECIPITATION, and DT_TEMPERATURE.

avdg81 committed:
Modified the unit test that verifies the LHS matrix calculation

Use a quadratic line condition rather than a linear one. The idea to is
have a test for each shape of condition (i.e. line, triangle, and
quadrilateral). At the same time, we would like to limit the tests to
quadratic shape functions only (to avoid an explosion of tests).

avdg81 committed:
Modified the unit test that verifies the RHS vector calculation

Use a quadratic line condition rather than a linear one. The idea to is
have a test for each shape of condition (i.e. line, triangle, and
quadrilateral). At the same time, we would like to limit the tests to
quadratic shape functions only (to avoid an explosion of tests).

avdg81 committed:
Added a test to verify the local system of a triangular condition

Only quadratic shape functions are tested.

To avoid repetition of test code, we now test both the LHS matrix and
the RHS vector that have been calculated in the same unit test. As a
result, two helper functions were no longer needed and they have been
removed.

Since we also need to compare a few computed zeros, we occasionally use
an absolute tolerance rather than a relative one.

avdg81 committed:
Added a test to verify the local system of a quadrilateral condition

avdg81 committed:
Renamed a data member to comply with the Kratos Style Guide

avdg81 committed:
Made a `public` data member `private`

To avoid having a mix of `public` and `private` data members, as was
found by SonarQube.

avdg81 committed:
Removed an unused data member

avdg81 committed:
Removed a needless `override`

The `override` did exactly the same thing as the base class
implementation.

avdg81 committed:
Removed a data member that was only assigned, but never used

Also removed two lines of commented out code.

avdg81 committed:
Inlined member function `CalculateKinematics`

avdg81 committed:
Removed some redundant comments, separation lines, and blank lines

avdg81 committed:
Improved several minor issues

- Removed some redundant blank lines.
- Ensure there is a space between `//` and the actual comment.
- Removed a redundant pair of parenthesis.
- Removed some commented out code.

avdg81 committed:
Improved the signature of a member function

- Return the result using the return value rather than using an out
  argument.
- Pass built-in types by value rather than by reference to const.

avdg81 committed:
Removed a redundant resize operation

Also removed a commented out declaration.

avdg81 committed:
Converted a data member to a local variable

Since the data member was only assigned to and used in a single function
body, it is better to use a local variable to express its intended
scope.

avdg81 committed:
Removed a needless assignment

avdg81 committed:
Converted a data member to a local variable

Since the data member was only assigned to and used in a single function
body, it is better to use a local variable to express its intended
scope.

avdg81 committed:
Removed a needless assignment

avdg81 committed:
Converted data member `previousRadiation` to a local variable

Since the data member was only assigned to and used in a single function
body, it is better to use a local variable to express its intended
scope.

avdg81 committed:
Removed the destructor that had a default implementation

When possible, try to apply the rule of zero.

avdg81 committed:
The `protected` members are now `private`

Since there no classes that inherit from `TMicroClimateFluxCondition`,
there is no point to have `protected` members.

Also, the `private` data members have been moved to the bottom of the
class declaration, which is a common place to put them.

avdg81 committed:
Removed an unused local variable

avdg81 committed:
Removed a needless resize operation

avdg81 committed:
No need to pass data members to `private` member functions

Removed a needless parameter from `private` member function
`InitializeElementVariables`.

Also removed a redundant comment.

avdg81 committed:
No need to pass data members to `private` member functions

Removed a needless parameter from `private` member function
`InitializeNodalTemperatureVariables`.

avdg81 committed:
No need to pass data members to `private` member functions

Removed a needless parameter from `private` member function
`CalculateAndAddRHS`.

Also removed a redundant comment.

avdg81 committed:
No need to pass data members to `private` member functions

Removed a needless parameter from `private` member function
`CalculateAndAddLHS`.

Also removed a redundant comment and a few redundant blank lines.

avdg81 committed:
No need to pass data members to `private` member functions

Removed a needless parameter from `private` member function
`CalculateRoughness`.

avdg81 committed:
No need to pass data members to `private` member functions

Removed a needless parameter from `private` member function
`CalculateRoughness`.

Also removed two redundant comments.

avdg81 committed:
Removed a redundant calculation of `TMatrix`

`TMatrix` was calculated in exactly the same way a few lines earlier.
Since none of its dependencies seem to have changed, the recalculation
is redundant.

avdg81 committed:
Added the integration tests from Mohamed's original branch

I had to make the following changes to let them pass:
- Remove the `"start_time"` property from `"solver_settings"`.
- Update the typename of the thermal elements by prepending it with
  `Geo`.
- Update the typename of the micro-climate flux conditions by prepending
  it with `Geo`.
- Replace material property `"SATURATION"` by `"SATURATION_LAW"` and
  `"SATURATED_SATURATION"`.

avdg81 committed:
Converted data member `TMatrix` to local variables

Since the scope of the data member was limited to two functions (and
there were no interdependencies), `TMatrix` has been replaced by two
local variables (both named `temporary_matrix`).

avdg81 committed:
Renamed local variables `TTMatrix` to `flux_matrix`

Mohamed proposed the new name.

avdg81 committed:
Converted data member `TVector` to a local variable

Since the scope of the data member was limited to a single function body
(and it was recalculated each time), we can use a local variable. In
this way, we are more explicit about the scope of the variable.

avdg81 committed:
Changed a fixed size array to an unbounded vector

avdg81 committed:
Added a utility function to get nodal values

This may help to simplify the code in many places.

avdg81 committed:
Removed unused data member `DtTemperatureVector`

This member was assigned, but never used.

Also removed a local variable that had become redundant. And removed a
comment that had no added value.

avdg81 committed:
Inlined member function `InitializeNodalTemperatureVariables`

avdg81 committed:
Replaced data member `TemperatureVector` by a local variable

Moved the retrieval of nodal temperatures one level up.

avdg81 committed:
Eliminated alias `VectorType`

It had no added value: it was neither clearer nor shorter.

avdg81 committed:
Eliminated alias `MatrixType`

It had no added value: it was neither clearer nor shorter.

avdg81 committed:
Member `GetNodalValues` now receives an output iterator

It no longer assumes a `Vector` as its result. Instead, the caller
provides an output iterator. Consequently, the result can be inserted
into any container, including fixed size arrays.

avdg81 committed:
Removed unused member `detJContainer`

It was only resized, but never used. Due to removing the resize
operation, another local variable became redundant.

avdg81 committed:
Converted member `NContainer` to a local variable

avdg81 committed:
Inlined member `InitializeElementVariables`

avdg81 committed:
Eliminated alias `NodeType`

avdg81 committed:
Eliminated alias `PropertiesType`

Also renamed a local variable to comply with the Kratos Style Guide.

avdg81 committed:
Cleaned up the constructors

- Removed some redundant comments.
- Removed some superfluous spaces.
- Broke a few long lines.
- Added spaces to separate template parameters.

avdg81 committed:
Cleaned up member `Create`

- Put each function parameter on its own line.
- Make sure that `const` is put in front of the typename.
- Renamed a function parameter.
- Use `make_intrusive` rather than a "naked" new.
- Make sure there is a space between template parameters.

avdg81 committed:
Use `MathUtils<>::Norm` to calculate the L2 norm of a vector

Removed a local reimplementation of this calculation.

avdg81 committed:
Added a helper function for creating `Vector` objects

This helps to simplify code that needs to create `Vector`s.

avdg81 committed:
Use `MathUtils<>::Norm` to calculate the L2 norm of a vector

Removed a local reimplementation of this calculation. To reuse
`MathUtils<>::Norm` the array type had to be changed to `Vector`.

avdg81 committed:
Use `MathUtils<>::CrossProduct` to calculate a cross product

Removed a local reimplementation of this calculation.

avdg81 committed:
Removed two needless resizing operations

Resetting the LHS matrix to a zero matrix has been reduced to a single
statement. Similarly, resetting the RHS vector to a zero vector has been
reduced to a single statement.

Also removed a redundant local variable.

avdg81 committed:
Extracted member `albedoCoefficient` from `ElementVariables`

It is now a data member of the condition class itself. Also eliminated a
local variable that had no additional value.

avdg81 committed:
Extracted member `firstCoverStorageCoefficient` from `ElementVariables`

It is now a data member of the condition class itself. Also eliminated a
local variable that had no additional value.

avdg81 committed:
Extracted member `secondCoverStorageCoefficient` from `ElementVariables`

It is now a data member of the condition class itself. Also eliminated a
local variable that had no additional value.

avdg81 committed:
Extracted member `secondCoverStorageCoefficient` from `ElementVariables`

It is now a data member of the condition class itself. Also eliminated a
local variable that had no additional value.

avdg81 committed:
Extracted member `buildEnvironmentRadiation` from `ElementVariables`

It is now a data member of the condition class itself. Also eliminated a
local variable that had no additional value.

avdg81 committed:
Extracted member `minimalStorage` from `ElementVariables`

It is now a data member of the condition class itself. Also eliminated a
local variable that had no additional value.

avdg81 committed:
Extracted member `maximalStorage` from `ElementVariables`

It is now a data member of the condition class itself. Also eliminated a
local variable that had no additional value.

avdg81 committed:
Extracted member `roughnessTemperature` from `ElementVariables`

It is now a data member of the condition class itself.

avdg81 committed:
Extracted member `netRadiation` from `ElementVariables`

It is now a data member of the condition class itself.

avdg81 committed:
Extracted member `waterStorage` from `ElementVariables`

It is now a data member of the condition class itself.

avdg81 committed:
Inlined member function `CalculateAll`

avdg81 committed:
A few minor improvements

- Removed a redundant pair of parenthesis.
- Removed several redundant spaces.

avdg81 committed:
Extracted a function that calculates the net radiation

Also replaced a few explicit typenames by `auto`.

avdg81 committed:
Renamed several local variables and function parameters

Made sure the names comply with the Kratos Style Guide.

avdg81 committed:
Renamed more local variables and a function parameter

Made sure the names comply with the Kratos Style Guide. Also replaced
some explicit typenames by `auto`.

Furthermore, an unused local variable has been removed. And another
variable was moved closer to its usage.

avdg81 committed:
Renamed more local variables and a function parameter

Made sure the names comply with the Kratos Style Guide. Also replaced
some explicit typenames by `auto`.

Furthermore, an assignment was taken out of the loop, since the assigned
value was always the same one.

avdg81 committed:
Several minor improvements

- Added a space to separate template parameters.
- Renamed a function parameter and several local variables to comply
  with the Kratos Style Guide. Also a few names have been made more
  accurate.
- Replaced some explicit typenames by `auto`.

avdg81 committed:
Reworked the calculations of the evaporation and precipitation

rfaasse committed:
Adding MicroClimateConstants namespace and split out setting the left hand side

rfaasse committed:
Split up the calculate nodal fluxes into smaller functions

avdg81 committed:
Added a helper function to create diagonal matrices

Replaced some duplicated code by calls to this new function.

rfaasse committed:
Clarified and renamed some functions

rfaasse committed:
Merge branch 'geo/11866-add-micro-climate-condition' of https://github.com/KratosMultiphysics/Kratos into geo/11866-add-micro-climate-condition

rfaasse committed:
Moved more constants to the micro_climate_constants.h

rfaasse committed:
Separate out setting the net radiation and left hand side fluxes

rfaasse committed:
Separate out the setting of water storage and right hand side fluxes

rfaasse committed:
Removed lefthandside flux from member variable, since it's only needed locally (and does not need history)

avdg81 committed:
Eliminated field `Np` from `ElementVariables`

The shape function values (`N`) are now passed to the functions that
need them through their argument lists.

avdg81 committed:
Eliminated field `IntegrationCoefficient` from `ElementVariables`

The integration coefficient is now passed to the functions that need it
through their argument lists.

rfaasse committed:
Moved righthandside vector out of the member struct since it was only needed locally

rfaasse committed:
Clarified the use of the net radiation/storage and removed function that became redundant

rfaasse committed:
Renamed the constants in micro_climate_constants.h to match the style guide

rfaasse committed:
Moving around functions and making relevant functions const

avdg81 committed:
Continued to reorder functions

Also removed a few redundant `this->` usages.

Furthermore, the entire `.cpp` file has been reformatted using
clang-format.

avdg81 committed:
Removed the remaining redundant `this->` usages

avdg81 committed:
Reformatted the entire `.hpp` file with clang-format

avdg81 committed:
Renamed `.hpp` file to `.h`

avdg81 committed:
Extracted a function to calculate the surface roughness factor

avdg81 committed:
Renamed a few local variables

Also added one more equation number.

rfaasse committed:
Fixed the initialization with an explanation and revalidated the UT results.

avdg81 committed:
Renamed several function parameters to comply with the style guide

avdg81 commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.h:
```diff
@@ -0,0 +1,152 @@
+};
+
+template <unsigned int TDim, unsigned int TNumNodes>
+class KRATOS_API(GEO_MECHANICS_APPLICATION) TMicroClimateFluxCondition
```
In line with the other geothermal condition classes (`GeoTCondition` and `GeoTNormalFluxCondition`), we may want to rename this class to `GeoTMicroClimateFluxCondition`.

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+template <unsigned int TDim, unsigned int TNumNodes>
+void TMicroClimateFluxCondition<TDim, TNumNodes>::Initialize(const ProcessInfo& rCurrentProcessInfo)
+{
+    Condition::Initialize(rCurrentProcessInfo);
```
Shouldn't this be `GeoTCondition::Initialize(rCurrentProcessInfo)`, since that is its direct base class?

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+    const auto flux_matrix = UBlasUtils::MakeDiagonalMatrix(
+        rLeftHandSideFluxes.begin(), rLeftHandSideFluxes.end());
+
+    temporary_matrix = prod(temporary_matrix, flux_matrix);
```
The creation of the diagonal matrix here is causing some extra numerical work.
If an elementwise vector multiplication is available, the shorter work can be expressed as:
outer_prod(rN, elementwiseMultiplication( rN, rLeftHandSideFluxes))

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void TMicroClimateFluxCondition<TDim, TNumNodes>::CalculateAndAddRHS(
```
From the coding I read that these are the integration point contributions for a flux based and a temperature based part. The flux and the temperature are given as element DOF vectors, which I do not think change per integration point. Therefore I wonder if setting up elementmatrices fully and then doing the matrix vector multiplications once i.s.o. number of integration point times is more efficient.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+    if (TDim == 2)
+    {
+        return MathUtils<>::Norm(
+                   UBlasUtils::MakeVector({rJacobian(0, 0), rJacobian(1, 0)})) *
```
Does the Matrix type have a method to get a column ( or row ). That would read easier than forming a vector of indicated elements of a matrix.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+        const auto vec1 =
+            UBlasUtils::MakeVector({rJacobian(0, 0), rJacobian(1, 0), rJacobian(2, 0)});
+        const auto vec2 =
+            UBlasUtils::MakeVector({rJacobian(0, 1), rJacobian(1, 1), rJacobian(2, 1)});
```
See previous comment, but I think that we have a very similiar implementation in the thermal element already that could be reused here.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+            AirHeatCapacity * AirDensity / RoughnessLayerResistance;
+        result[i] = sensible_heat_flux_left;
+    }
+
```
an array filled with same values? Then compute the value once and don't program a loop.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+        const auto actual_storage =
+            PreviousStorage +
+            TimeStepSize * (water_fluxes.precipitation - water_fluxes.evaporation);
+        result += actual_storage / TNumNodes;
```
The division can be done once, outside the loop.

rfaasse replied:
Done

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+    for (unsigned int i = 0; i < TNumNodes; ++i)
+    {
+        const auto net_radiation = CalculateNetRadiation(i);
+        result += net_radiation / TNumNodes;
```
The division can be outside the loop. Then the loop reduces to a summation of CalculateNetRadiation(i)

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+        r_geom[NodeIndex].FastGetSolutionStepValue(SOLAR_RADIATION);
+    const auto short_wave_radiation = (1.0 - mAlbedoCoefficient) * incoming_radiation;
+
+    // Eq 5.17
+    const auto atmospheric_temperature =
+        r_geom[NodeIndex].FastGetSolutionStepValue(AIR_TEMPERATURE);
+    const auto absorbed_long_wave_radiation =
+        EffectiveEmissivity * BoltzmannCoefficient *
+        std::pow(atmospheric_temperature + 273.15, 4.0);
+
+    // Eq 5.18
+    const auto initial_soil_temperature =
+        r_geom[NodeIndex].FastGetSolutionStepValue(TEMPERATURE, 1);
+    const auto emitted_long_wave_radiation =
+        BoltzmannCoefficient * std::pow(initial_soil_temperature + 273.15, 4.0);
```
This seems to assume that temperatures are given in Celsius. It also assumes that the far field temperature is 0 Kelvin. ( when the formula is BoltzmannCoefficient * (T_surface^4 - T_farfield^4)

rfaasse replied:
I added to the documentation that this assumption should be documented

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+    auto actual_evaporation = potential_evaporation;
+
+    // Correct the precipitation and evaporation if the storage is out of bounds
+    if (potential_storage > mMaximalStorage)
+    {
+        actual_precipitation =
+            (mMaximalStorage - PreviousStorage) / TimeStepSize + actual_evaporation;
+    }
+    else if (potential_storage < mMinimalStorage)
+    {
+        actual_evaporation = (PreviousStorage - mMinimalStorage) / TimeStepSize + actual_precipitation;
+    }
```
Please use min and max.

rfaasse replied:
I have taken some time to try and see if this is possible, but at this moment, I really don't see how. The problem is that the if-statements will remain necessary, because depending on if they're true, we either correct precipitation or evaporation. And the statements themselves are also not trivial to put into std::min/max format. If you have a good suggestion, please let me know! Or maybe @avdg81 sees a possibility?

avdg81 replied:
No, I also don't see how we could do that. Therefore, I'd suggest to keep it the way it is now. Unless @WPK4FEM has an elegant solution.

WPK4FEM replied:
I do not see an easy way. Please ignore my initial remark.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+    WaterFluxes water_fluxes;
+    water_fluxes.evaporation = actual_evaporation;
+    water_fluxes.precipitation = actual_precipitation;
+    return water_fluxes;
```
Can a constructor that gives a value to evaporation and precipitation be used?

rfaasse replied:
Simplified, good suggestion!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+
+    // Eq 5.35
+    const auto wind_speed = r_geom[NodeIndex].FastGetSolutionStepValue(WIND_SPEED);
+    const auto atmospheric_resistance = 1.0 / (0.007 + 0.0056 * wind_speed);
```
What is the assumed unit for wind speed?

rfaasse replied:
I assume m/s, I added to the documentation issue that it should be specified

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+
+    // Eq 5.14
+    const auto humidity = r_geom[NodeIndex].FastGetSolutionStepValue(AIR_HUMIDITY);
+    const auto actual_vapor_pressure = humidity / 100.0 * saturated_vapor_pressure;
```
Then humidity is between 0 and 100% ?
For human readers the formula becomes ambiguous ( humidity /100 ) * saturated_vapor_pressure or humidity / ( 100 * saturated_vapor_pressure )

rfaasse replied:
I guess so, I'll add it to the issue for documentation of microclimate that it should be documented humidity is in %. Suggestion is also done!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+                 surface_roughness_factor * friction_drag_coefficient *
+                 friction_drag_coefficient * current_air_temperature) /
+            denominator;
+        mRoughnessTemperature += current_roughness_temperature / TNumNodes;
```
Division outside the loop?

rfaasse replied:
Done

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+            richardson_bulk_modulus, friction_drag_coefficient);
+
+        // Eq 5.30
+        const auto denominator =
+            RoughnessLayerResistance * RoughnessLayerHeight + time_step_size +
+            time_step_size * current_wind_speed * RoughnessLayerResistance *
+                surface_roughness_factor * friction_drag_coefficient * friction_drag_coefficient;
+        const auto current_roughness_temperature =
+            (RoughnessLayerResistance * RoughnessLayerHeight * previous_roughness_temperature +
+             time_step_size * initial_soil_temperature +
+             time_step_size * current_wind_speed * RoughnessLayerResistance *
+                 surface_roughness_factor * friction_drag_coefficient *
+                 friction_drag_coefficient * current_air_temperature) /
+            denominator;
```
Does this give consistent units?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/micro_climate_constants.h:
```diff
@@ -0,0 +1,34 @@
+constexpr static auto AirHeatCapacity = 1004.67;
+constexpr static auto RoughnessLayerResistance = 30.0;
+constexpr static auto LatentEvaporationHeat = 2.45e6;
+constexpr static auto WaterDensity = 1e3;
```
Redefinition, this already exists in our material input and does vary there. Furthermore it dictates what the system of units should be.

rfaasse replied:
Good catch! I removed it from here and used the DENSITY_WATER parameter from the material parameters in the condition

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/micro_climate_constants.h:
```diff
@@ -0,0 +1,34 @@
+constexpr static auto VonNeumannCoefficient = 0.4;
+constexpr static auto MeasurementHeight = 10.0;
+constexpr static auto RoughnessHeight = 1.0;
+constexpr static auto GravitationalAcceleration = 9.81;
```
This is probably the 4th redefinition of g that we have ( volumetric acceleration process, weights in waterloads etc. ) It really should be only 1 and 1 alone!

rfaasse replied:
Completely agreed, however we would like to fix this for all places at once, instead of only for this one (and it's not clear right now where that 1 place would be yet)

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/micro_climate_constants.h:
```diff
@@ -0,0 +1,34 @@
+//
+#pragma once
+
+namespace Kratos::MicroClimateConstants
```
Wise to have 1 place for definitions.

They seem a mix of material parameters and physical constants. Sadly some of them are redefinitions of already existing parameters.

rfaasse replied:
We removed the redefinition of the water density and as seen in the other comment, the gravity constant is a bigger issue that we'd like to tackle seperately

WPK4FEM commented on applications/GeoMechanicsApplication/custom_python/geo_mechanics_python_application.cpp:
```diff
@@ -62,6 +62,12 @@ PYBIND11_MODULE(KratosGeoMechanicsApplication,m)
 
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, DT_TEMPERATURE)
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, NORMAL_HEAT_FLUX)
+	KRATOS_REGISTER_IN_PYTHON_VARIABLE( m, AIR_TEMPERATURE )
```
Indentational difference.
Should the registrations always be there or can they be conditional?

rfaasse replied:
Replaced a tab with spaces. 

I would propose to look at conditional registration of variables in a new issue. I think it is a bit out of scope, since we need to figure out how to do that properly (in a way that it also works for other conditionals)

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py:
```diff
@@ -361,6 +361,11 @@ def _add_temperature_variables(self):
+        self.main_model_part.AddNodalSolutionStepVariable(GeoMechanicsApplication.SOLAR_RADIATION)
+        self.main_model_part.AddNodalSolutionStepVariable(GeoMechanicsApplication.AIR_HUMIDITY)
+        self.main_model_part.AddNodalSolutionStepVariable(GeoMechanicsApplication.PRECIPITATION)
+        self.main_model_part.AddNodalSolutionStepVariable(GeoMechanicsApplication.WIND_SPEED)        
```
Why the difference in indentation?

rfaasse replied:
For me it looks okay, maybe it's a github issue?
![image](https://github.com/KratosMultiphysics/Kratos/assets/56549273/d74dbf9c-affb-4537-ba3d-5c780fa9c1df)

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_micro_climate_1/MaterialParameters.json:
```diff
@@ -0,0 +1,36 @@
+        "QF_COEFFICIENT"                   : 0.000000e+00,
+        "SMIN_COEFFICIENT"                 : 0.000000e+00,
+        "SMAX_COEFFICIENT"                 : 4.800000e-04
+      },
```
These are new and should now appear on our wiki.

rfaasse replied:
I added to the documentation issue, we should add this to the wiki

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_micro_climate_1/ProjectParameters.json:
```diff
@@ -0,0 +1,209 @@
+      }
+    }]
+  },
+  "processes": {
```
the processes for new variables deserve a place on our wiki.

rfaasse replied:
I added to the documentation issue, we should add this to the wiki

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_micro_climate_2/ProjectParameters.json:
```diff
@@ -0,0 +1,209 @@
+            "skin_output"                : false,
+            "plane_output"               : [],
+            "nodal_results"              : ["TEMPERATURE"],
+            "gauss_point_results"        : []
```
I have not seen HEAT_FLUX output yet. It would be a natural thing to have, like we also have fluid flux.

rfaasse replied:
I created a draft issue on our product backlog to do this

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_micro_climate_2/ProjectParameters.json:
```diff
@@ -0,0 +1,209 @@
+      "process_name"                     : "ApplyScalarConstraintTableProcess",
+      "Parameters" : {
+        "model_part_name"                : "PorousDomain.TopThermo",
+        "variable_name"                  : "AIR_TEMPERATURE",
```
can we output and display these new variables? that would be a natural addition for model checking.

rfaasse replied:
I'm not sure, I created a draft issue on our product backlog to check/implement this

rfaasse committed:
Removed trailing white spaces (codacy)

WPK4FEM reviewed (commented):
Impressive testing!
The restructuring probably made clear that some more improvements can be made.
This implementation relies on input being in a precisely defined set of units that hardly can be checked with input verification. The redefinition of already existing material parameters is dangerous.

markelov208 commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,483 @@
+    const auto& r_geom = this->GetGeometry();
+    const auto current_air_temperature = r_geom[0].FastGetSolutionStepValue(AIR_TEMPERATURE);
+    const auto current_wind_speed =
+        std::max(1.0e-3, r_geom[0].FastGetSolutionStepValue(WIND_SPEED));
```
Is it better to have a named constant for 1.0e-3?

rfaasse replied:
Agreed, renamed it

markelov208 commented on applications/GeoMechanicsApplication/custom_python/geo_mechanics_python_application.cpp:
```diff
@@ -62,6 +62,12 @@ PYBIND11_MODULE(KratosGeoMechanicsApplication,m)
 
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, DT_TEMPERATURE)
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, NORMAL_HEAT_FLUX)
+	KRATOS_REGISTER_IN_PYTHON_VARIABLE( m, AIR_TEMPERATURE )
```
Is it indentation?

rfaasse replied:
Replaced a tab with spaces

markelov208 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_T_micro_climate_flux_condition.cpp:
```diff
@@ -0,0 +1,437 @@
+
+namespace Kratos::Testing {
+
+KRATOS_TEST_CASE_IN_SUITE(NoThrowWhenInitializingThermalMicroClimateCondition, KratosGeoMechanicsFastSuite)
```
Due to the great generalization, the tests look almost the same. Is it possible to combine them? Or will it be done later with gtest?

rfaasse replied:
The scope of the tests is still limited and in this way they are easy to read independently, so for now I'd propose to keep this as it is.

markelov208 reviewed (commented):
Richard, very well done. Just a couple of comments.

avdg81 committed:
Moved the calculation of the integration coefficient

It is now part of the condition utilities. As a result, the newly
introduced function `MakeVector` had become redundant, which has been
removed now.

avdg81 committed:
Reuse the function to calculate the integration coefficient

avdg81 committed:
Use `elemen_prod` for elementwise vector multiplications

As a result, we no longer need the newly introduced function
`MakeDiagonalMatrix`.

avdg81 committed:
Simplified a function's implementation

Eliminated a "raw" loop by selecting a more appropriate constructor.

rfaasse committed:
Rewrote summing the net radiation using std::accumulate

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,448 @@
+
+    // Containers of variables at all integration points
+    const auto local_dim = static_cast<unsigned int>(r_geom.LocalSpaceDimension());
+    GeometryType::JacobiansType J_container(number_of_integration_points);
```
To adhere to the style guide, this should be j_container

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,448 @@
+    for (unsigned int integration_point_index = 0;
+         integration_point_index < number_of_integration_points; ++integration_point_index)
+    {
+        const auto N =
```
Maybe we can think of a better name for N?

WPK4FEM replied:
In finite elements N is the general notation for the interpolation polynomial, it matches with almost all textbooks. Therefore please let N remain N.

rfaasse replied:
Okay, clear we'll keep it like it is

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/micro_climate_constants.h:
```diff
@@ -0,0 +1,34 @@
+namespace Kratos::MicroClimateConstants
+{
+
+constexpr static auto AirDensity = 1.18;
```
I now see (next to WPKs comments) that there might be a better way to sort these constants

rfaasse commented on applications/GeoMechanicsApplication/geo_mechanics_application.h:
```diff
@@ -535,6 +536,17 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) KratosGeoMechanicsApplication : publ
+    const TMicroClimateFluxCondition<3, 8> mGeoTMicroClimateFluxCondition3D8N{0, Kratos::make_shared<Quadrilateral3D8<NodeType>>(Condition::GeometryType::PointsArrayType(8))};
+    const TMicroClimateFluxCondition<3, 9> mGeoTMicroClimateFluxCondition3D9N{0, Kratos::make_shared<Quadrilateral3D9<NodeType>>(Condition::GeometryType::PointsArrayType(9))};
+
+
```
We can remove this unnecessary white line

rfaasse commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py:
```diff
@@ -361,6 +361,11 @@ def _add_temperature_variables(self):
         self.main_model_part.AddNodalSolutionStepVariable(GeoMechanicsApplication.DT_TEMPERATURE)
         # Add variables for the heat conditions
         self.main_model_part.AddNodalSolutionStepVariable(GeoMechanicsApplication.NORMAL_HEAT_FLUX)
+        self.main_model_part.AddNodalSolutionStepVariable(GeoMechanicsApplication.AIR_TEMPERATURE)
```
Let's make an issue to make these additions conditional (e.g. only when we use the t_solver?)

rfaasse reviewed (commented):
Only a few minor comments left when going through the code a last time.

avdg81 committed:
Removed two unused material parameters from the new input files

Parameters `LONGITUDINAL_DISPERSIVITY` and `TRANSVERSE_DISPERSIVITY` are
not being used by the geomechanical code at all. Therefore, those have
been removed from the new input files for integration testing.

rfaasse committed:
Processing part of the review comments

rfaasse committed:
Implemented save/load for the microclimate condition members

rfaasse committed:
Reading water density from file instead of setting it a constant

avdg81 committed:
Fixed the calculation of the current net radiation

For some reason, in the previous implementation, each node was
implicitly converted to the number 1 (WTF?). Now we use the `std::iota`
algorithm to generate a sequence of local node indices and feed that to
the `std::accumulate` algorithm.

avdg81 committed:
Renamed a local variable

avdg81 committed:
Prepended the micro climate condition class name with `Geo`

avdg81 committed:
Corrected the qualification of the `Initialize` call

It used to call the member function of an indirect base class. Now it
calls the one from its direct base class.

avdg81 committed:
Replaced a <Tab> by spaces

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/11866-add-micro-climate-condition

markelov208 commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,452 @@
+            array_1d<double, TNumNodes>{row(r_N_container, integration_point_index)};
+
+        // Compute weighting coefficient for integration
+        const auto IntegrationCoefficient =
```
should it be integration_coefficient?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,452 @@
+    // Eq 5.31
+    const auto subsurface_heat_flux =
+        NetRadiation - sensible_heat_flux_right - latent_heat_flux +
+        mBuildEnvironmentRadiation - SurfaceHeatStorage;
```
misspelled? built environment

avdg81 committed:
Fixed a compile-time error from GCC and Clang

Don't forget to specify the template arguments.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -0,0 +1,452 @@
+        r_geom[NodeIndex].FastGetSolutionStepValue(AIR_TEMPERATURE);
+    const auto absorbed_long_wave_radiation =
+        EffectiveEmissivity * BoltzmannCoefficient *
+        std::pow(atmospheric_temperature + 273.15, 4.0);
```
The 273.15 shift from Celsius to Kelvin may deserve to be a named constant.

WPK4FEM reviewed (approved):
I think it is good to do the remaining things after consultation with Mohamed. Lets go ahead merging the current status.

markelov208 reviewed (commented):
Anne and Richard, great work. Just one comment on the variable name.
